### PR TITLE
Generate report by TestTypeId

### PR DESCRIPTION
**Pull request by harsinha**

### CODE REVIEW

File: src/application/service/reports/GetReports.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { TestTypeId } from '../../../domain/model/test/testType/TestTypeId';
+import { ReportRepository } from '../../../domain/model/reports/ReportRepository';
+import { ReportTestResult } from '../../../domain/model/reports/ReportTestResult';
+
+export class GetReports {
+  constructor(private reportRepository: ReportRepository) {}
+
+  async getTestResults(testTypeId: TestTypeId): Promise<Array<ReportTestResult>> {
```
Comment (kstam) [inline src/application/service/reports/GetReports.ts:8]: Neatpicking: I would call this byTestTypeId to follow the same convention as other `Get*` classes

File: src/infrastructure/persistence/PsqlReportRepository.test.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+describe('PsqlReportRepository', () => {
+  const psqlReportRepository = new PsqlReportRepository(database, testTypeRepository);
+  const psqlTestRepository = new PsqlTestRepository(database, testTypeRepository);
+
+  beforeEach(async () => {
+    await cleanupDatabase();
+  });
+
+  it('inserts a test without results and gets report', async () => {
```
Comment (kstam) [inline src/infrastructure/persistence/PsqlReportRepository.test.ts:23]: I guess here we aren't testing the insert. Maybe this should read: "gets results for a single test"?

File: src/infrastructure/persistence/PsqlReportRepository.test.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    const reportTestResults = await psqlReportRepository.getTestResults(testType.id);
+
+    expect(reportTestResults[0].id).toEqual(test.id);
+    expect(reportTestResults[0].userId).toEqual(test.userId);
+    expect(reportTestResults[0].testType.id).toEqual(test.testType.id);
+    expect(reportTestResults[0].resultsDetails).toBeNull();
+  });
+
+  it('inserts 2 tests with results and gets report', async () => {
```
Comment (kstam) [inline src/infrastructure/persistence/PsqlReportRepository.test.ts:37]: same here: "gets multiple test results"

File: src/infrastructure/persistence/PsqlReportRepository.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  `${TEST_TABLE_NAME}.user_id as userId`,
+  `${TEST_TABLE_NAME}.test_type_id as testTypeId`,
+  `${TEST_TABLE_NAME}.creation_time as testCreationTime`,
+  `${TEST_TABLE_NAME}.creator_user_id as testCreatorUserId`,
+  `${TEST_TABLE_NAME}.creator_confidence as testCreatorConfidence`,
+  `${TEST_RESULTS_TABLE_NAME}.creator_user_id as resultsCreatorUserId`,
+  `${TEST_RESULTS_TABLE_NAME}.details as resultsDetails`,
+  `${TEST_RESULTS_TABLE_NAME}.notes as resultsNotes`,
+  `${TEST_RESULTS_TABLE_NAME}.creator_confidence as resultsCreatorConfidence`,
+  `${TEST_RESULTS_TABLE_NAME}.creation_time as resultsCreationTime`,
```
Comment (kstam) [inline src/infrastructure/persistence/PsqlReportRepository.ts:25]: Didn't we also want to extract the user's address (Postcode, city etc)?

File: src/presentation/api/reports/ReportsController.test.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      const testTypeIdValue = testType.id.value;
+
+      await request(app)
+        .get(`/api/v1/reports/test-results/?testTypeId=${testTypeIdValue}`)
+        .set({
+          Authorization: `Bearer ${await getTokenForUser(authenticatedUser)}`,
+        })
+        .expect(200)
+        .expect((res) => {
+          //          expect(res.body[0].).toEqual('mail@example.com');
```
Comment (kstam) [inline src/presentation/api/reports/ReportsController.test.ts:35]: THis test isn't actually asserting anything

File: src/presentation/api/reports/ReportsController.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+import { writeToStream } from '@fast-csv/format';
+
+@Authorized()
+@JsonController('/v1/reports')
+export class ReportsController {
+  private getReports = getReports;
+  private getTestTypes = getTestTypes;
+
+  // Only used for data reporting/export in MVP. Very heavy query as database grows.
```
Comment (kstam) [inline src/presentation/api/reports/ReportsController.ts:17]: Is there something we can change in the name of the method to make this comment not needed? 😇

File: src/presentation/api/reports/ReportsController.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  @Get('/test-results')
+  @UseBefore(hasPermission(VIEW_ADMIN_REPORTS))
+  async getReportByTestTypeId(@QueryParam('testTypeId') testTypeId: string, @Req() req: Request, @Res() res: Response) {
+    const testType = await this.getTestTypes.byId(testTypeId);
+
+    if (!testType) {
+      throw new ApiError(404, apiErrorCodes.TEST_TYPE_NOT_FOUND);
+    }
+
+    const tests = await this.getReports.getTestResults(testType.id);
```
Comment (kstam) [inline src/presentation/api/reports/ReportsController.ts:27]: Maybe not for now but would it make sense to add a date to the request parameters and return only results after a certain date?

File: src/presentation/transformers/reports/transformReportTestResultsToReport.test.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import { aReportTestResult, aResult, aTestType } from '../../../test/domainFactories';
+import { transformReportTestResultsToReport } from './transformReportTestResultsToReport';
+import { ReportTestResult } from '../../../domain/model/reports/ReportTestResult';
+
+describe('transformReportTestResultsToReport', () => {
+  it('correctly transforms a ReportTestResult to Report data', () => {
```
Comment (kstam) [inline src/presentation/transformers/reports/transformReportTestResultsToReport.test.ts:6]: missing test for dynamic test result fields

Comment (harsinha) [reply src/presentation/transformers/reports/transformReportTestResultsToReport.test.ts:6]: Good catch. 👍 Actually there was a logic bug in the test data set up here too.

File: src/infrastructure/persistence/PsqlReportRepository.ts (hunk 4 above)
Comment (kstam) [reply src/infrastructure/persistence/PsqlReportRepository.ts:25]: Agreed we'd park for now

**Pull request merged**